operator/auth/spire/client.go

```diff
@@ -0,0 +1,247 @@
+		"The trust domain for the SPIFFE identity.")
+}
+
+type Client struct {
+	cfg   ClientConfig
+	log   logrus.FieldLogger
+	entry entryv1.EntryClient
+}
+
+// NewClient creates a new SPIRE client.
+// If the mTLS is not enabled, it returns a noop client.
+func NewClient(lc hive.Lifecycle, cfg ClientConfig, log logrus.FieldLogger) identity.Provider {
+	if !cfg.AuthMTLSEnabled {
+		return &noopClient{}
+	}
+	client := &Client{
+		cfg: cfg,
+		log: log.WithField(logfields.LogSubsys, "spire-client"),
+	}
+
+	lc.Append(hive.Hook{
+		OnStart: client.onStart,
+		OnStop:  func(_ hive.HookContext) error { return nil },
+	})
+	return client
+}
+
+func (c *Client) onStart(_ hive.HookContext) error {
+	go func() {
+		c.log.Info("Initializing SPIRE client")
+		attempts := 0
+		backoffTime := backoff.Exponential{Min: 100 * time.Millisecond, Max: 10 * time.Second}
+		for {
+			attempts++
+			conn, err := c.connect(context.Background())
+			if err == nil {
+				c.entry = entryv1.NewEntryClient(conn)
+				break
+			}
+			c.log.WithError(err).Errorf("Unable to connect to SPIRE server, attempt %d", attempts+1)
+			time.Sleep(backoffTime.Duration(attempts))
+		}
+		c.log.Info("Initialized SPIRE client")
+	}()
+	return nil
+}
+
+func (c *Client) connect(ctx context.Context) (*grpc.ClientConn, error) {
+	timeoutCtx, cancelFunc := context.WithTimeout(ctx, c.cfg.SpireServerConnectionTimeout)
+	defer cancelFunc()
+
+	// This is blocking till the cilium-operator is registered in SPIRE.
+	source, err := workloadapi.NewX509Source(timeoutCtx,
+		workloadapi.WithClientOptions(
+			workloadapi.WithAddr(fmt.Sprintf("unix://%s", c.cfg.SpireAgentSocketPath)),
+			workloadapi.WithLogger(c.log),
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create X509 source: %w", err)
+	}
+
+	trustedDomain, err := spiffeid.TrustDomainFromString(c.cfg.SpiffeTrustDomain)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse trust domain: %w", err)
+	}
+
+	tlsConfig := tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeMemberOf(trustedDomain))
+	conn, err := grpc.Dial(c.cfg.SpireServerAddress, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create connection to SPIRE server: %w", err)
+	}
+
+	return conn, nil
+}
+
+// Upsert creates or updates the SPIFFE ID for the given ID.
+// The SPIFFE ID is in the form of spiffe://<trust-domain>/cilium-id/<id>.
+func (c *Client) Upsert(ctx context.Context, id string) error {
+	if c.entry == nil {
+		return initializedError
+	}
+
+	entries, err := c.listEntries(ctx, id)
+	if err != nil && !strings.Contains(err.Error(), notFoundError) {
+		return err
+	}
+
+	desired := []*types.Entry{
+		{
+			SpiffeId: &types.SPIFFEID{
+				TrustDomain: c.cfg.SpiffeTrustDomain,
+				Path:        toPath(id),
+			},
+			ParentId: &types.SPIFFEID{
+				TrustDomain: c.cfg.SpiffeTrustDomain,
+				Path:        defaultParentID,
+			},
+			Selectors: defaultSelectors,
```

Comment:
It's possible we may need to make these configurable in the future (more likely the `defaultParentID` than the `defaultSelectors` though.)